pkg/antctl/antctl.go

```diff
@@ -0,0 +1,140 @@
+	"github.com/vmware-tanzu/antrea/pkg/version"
+)
+
+// unixDomainSockAddr is the address for antctl server in local mode.
+const unixDomainSockAddr = "/var/run/antctl.sock"
+
+var singleton sync.Once
+
+type versionResponse struct {
+	handlers.ComponentVersionResponse `json:",inline" yaml:",inline"`
+	AntctlVersion                     string `json:"antctlVersion" yaml:"antctlVersion"`
+}
+
+func versionTransform(reader io.Reader, _ bool) (interface{}, error) {
+	b, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+	klog.Infof("version transform received: %s", string(b))
+	var cv handlers.ComponentVersionResponse
+	err = json.Unmarshal(b, &cv)
+	if err != nil {
+		return nil, err
+	}
+	resp := &versionResponse{
+		ComponentVersionResponse: cv,
+		AntctlVersion:            version.GetFullVersion(),
+	}
+	return resp, nil
+}
+
+// CommandList defines all commands could be used for the antctl for both agent
+// and controller. The unit test TestManifestValidation ensures it to be valid.
+var CommandList = &commandList{
+	definitions: []commandDefinition{
+		{
+			Use:            "version",
+			Short:          "Print version information",
+			Long:           "Print version information of the antctl and the ${component}",
+			HandlerFactory: new(handlers.Version),
+			ResponseStruct: new(versionResponse),
+			Agent:          true,
+			Controller:     true,
+			Singleton:      true,
+			CommandGroup:   flat,
+			AddonTransform: versionTransform,
+		},
+	},
+	groupVersion: &schema.GroupVersion{Group: "clusterinfo.antrea.tanzu.vmware.com", Version: "v1"},
+	codec:        scheme.Codecs,
+}
+
+// Server describes actions of an antctl server.
+type Server interface {
+	// Run starts the antctl server, implementations need to use the value of AgentMonitor
+	// and Controller monitor to know which component it runs in. It should return
+	// a read-only channel which is used to tell whether the server is stopped.
+	Run(aq monitor.AgentQuerier, cq monitor.ControllerQuerier, stopCh <-chan struct{}) <-chan struct{}
+}
+
+type server struct {
+	runOnce   sync.Once
+	listener  net.Listener
+	stoppedCh chan struct{}
+}
+
+func (s *server) Run(aq monitor.AgentQuerier, cq monitor.ControllerQuerier, stopCh <-chan struct{}) <-chan struct{} {
+	s.runOnce.Do(func() {
+		antctlMux := mux.NewPathRecorderMux("antctl-server")
+		CommandList.ApplyToMux(antctlMux, aq, cq)
+		server := &http.Server{Handler: antctlMux}
+		// Request http server graceful stop
+		go func() {
+			<-stopCh
+			err := server.Shutdown(context.Background())
+			if err != nil {
+				klog.Errorf("Antctl server stopped with error: %v", err)
+			} else {
+				klog.Info("Antctl server stopped")
+			}
+		}()
+		// Run the http server
+		go func() {
+			klog.Info("Starting antctl server")
+			err := server.Serve(s.listener)
+			if err != nil {
+				klog.Errorf("Antctl server stopped with error: %v", err)
+			}
+			close(s.stoppedCh)
+		}()
+	})
+	return s.stoppedCh
```

Comment:
Not sure if I miss anything here, but it seems that the return channel is never used? Then why we need to introduce this channel?

Reply:
Thanks, removed.